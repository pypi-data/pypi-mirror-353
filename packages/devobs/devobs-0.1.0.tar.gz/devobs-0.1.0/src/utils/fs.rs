use std::{fs::create_dir_all,
          path::{Path, PathBuf}};

use anyhow::Result;
use glob::glob;

/// Create the file if it does not exist, including its parent directories.
pub(crate) fn touch_file(path: &Path) -> Result<()> {
    if path.exists() {
        log::debug!("File already exists: {}", path.display());
        return Ok(());
    }

    create_dir_all(
        path.parent()
            .expect("Failed to get parent directory for file creation."),
    )?;
    std::fs::File::create(path)?;
    log::debug!("Created file: {}", path.display());

    Ok(())
}

/// List files in the `from` directory based on the include and exclude patterns.
pub(crate) fn list_files(from: &Path, include: &[String], exclude: &[String]) -> Vec<PathBuf> {
    let mut include = expand_glob(from, include);
    let exclude = expand_glob(from, exclude);

    // Filter out files that match the exclude patterns
    include.retain(|path| {
        // Exclude files that match any of the exclude patterns
        !exclude.iter().any(|ex| path == ex)
    });

    include
}

/// Expand glob patterns in the given directory, returning a flat list of paths.
pub(crate) fn expand_glob(from: &Path, patterns: &[String]) -> Vec<PathBuf> {
    patterns
        .iter()
        .flat_map(|s| {
            glob(
                from.join(s)
                    .to_str()
                    .expect("Failed to convert path to string"),
            )
            .expect("Failed to create glob pattern")
        })
        .filter_map(Result::ok)
        .collect()
}

#[cfg(test)]
mod tests {
    use anyhow::Result;
    use sugars::hmap;

    use super::*;
    use crate::helpers::get_temp_dir;

    #[test]
    fn test_touch_file() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {});
        let file_path = temp_dir.path().join("test.txt");
        assert!(!file_path.exists());

        // Act
        touch_file(&file_path)?;

        // Assert
        assert!(file_path.exists());
        Ok(())
    }

    #[test]
    fn test_touch_file_nested_directory() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {});
        let nested_file_path = temp_dir.path().join("nested/dir/test.txt");
        assert!(!nested_file_path.exists());

        // Act
        touch_file(&nested_file_path)?;

        // Assert
        assert!(nested_file_path.exists());
        Ok(())
    }

    #[test]
    fn test_touch_existing_file() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {"test.txt" => ""});
        let file_path = temp_dir.path().join("test.txt");
        assert!(file_path.exists());

        // Act
        touch_file(&file_path)?;

        // Assert
        assert!(file_path.exists());
        Ok(())
    }

    #[test]
    fn test_expand_glob_simple() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {
            "file1.txt" => "",
            "file2.txt" => "",
            "other.log" => "",
        });
        let dir_path = temp_dir.path();

        // Act
        let mut txt_files = expand_glob(dir_path, &["*.txt".to_string()]);
        txt_files.sort();

        // Assert
        assert_eq!(
            txt_files,
            &[dir_path.join("file1.txt"), dir_path.join("file2.txt")]
        );
        Ok(())
    }

    #[test]
    fn test_expand_glob_multiple_patterns() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {
            "file1.txt" => "",
            "file2.txt" => "",
            "other.log" => "",
        });
        let dir_path = temp_dir.path();

        // Act
        let mut all_files = expand_glob(dir_path, &["*.txt".to_string(), "*.log".to_string()]);
        all_files.sort();

        // Assert
        assert_eq!(
            all_files,
            &[
                dir_path.join("file1.txt"),
                dir_path.join("file2.txt"),
                dir_path.join("other.log"),
            ]
        );
        Ok(())
    }

    #[test]
    fn test_expand_glob_recursive() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {
            "file1.txt" => "",
            "file2.txt" => "",
            "subdir/nested.txt" => "",
        });
        let dir_path = temp_dir.path();

        // Act
        let mut all_nested = expand_glob(dir_path, &["**/*.txt".to_string()]);
        all_nested.sort();

        // Assert
        assert_eq!(
            all_nested,
            &[
                dir_path.join("file1.txt"),
                dir_path.join("file2.txt"),
                dir_path.join("subdir/nested.txt"),
            ]
        );
        Ok(())
    }

    #[test]
    fn test_list_files_with_exclude() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {
            "file1.txt" => "",
            "file2.txt" => "",
            "file3.txt" => "",
        });
        let dir_path = temp_dir.path();

        // Act
        let mut files = list_files(dir_path, &["*.txt".to_string()], &["file2.txt".to_string()]);
        files.sort();

        // Assert
        assert_eq!(
            files,
            &[dir_path.join("file1.txt"), dir_path.join("file3.txt")]
        );
        Ok(())
    }

    #[test]
    fn test_list_files_recursive_with_exclude() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {
            "file1.txt" => "",
            "file2.txt" => "",
            "file3.txt" => "",
            "other.log" => "",
            "subdir/nested.txt" => "",
            "subdir/nested.log" => "",
        });
        let dir_path = temp_dir.path();

        // Act
        let mut files = list_files(
            dir_path,
            &["**/*.txt".to_string()],
            &["**/*.log".to_string()],
        );
        files.sort();

        // Assert
        assert_eq!(
            files,
            &[
                dir_path.join("file1.txt"),
                dir_path.join("file2.txt"),
                dir_path.join("file3.txt"),
                dir_path.join("subdir/nested.txt"),
            ]
        );
        Ok(())
    }

    #[test]
    fn test_list_files_empty_patterns() -> Result<()> {
        // Arrange
        let temp_dir = get_temp_dir(hmap! {
            "file1.txt" => "",
            "file2.txt" => "",
        });
        let dir_path = temp_dir.path();

        // Act
        let files = list_files(dir_path, &[], &[]);

        // Assert
        assert_eq!(files, &[] as &[PathBuf]);
        Ok(())
    }
}
