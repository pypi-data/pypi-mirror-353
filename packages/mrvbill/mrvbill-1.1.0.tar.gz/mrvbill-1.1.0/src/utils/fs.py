from pathlib import Path
import json

# Gets the config file
def get_config_file() -> Path:
    """Gets the config file path"""
    config_dir = Path.home() / '.pybill'
    config_dir.mkdir(exist_ok=True)  # Create directory if it doesn't exist
    return config_dir / 'config.json'

# Writes the value to the config file
def write_to_config(key: str, value: str) -> None:
    """
    Writes a key-value pair to the config file.
    Creates the file and directory if they don't exist.
    """
    config_file = get_config_file()
    config_file.parent.mkdir(parents=True, exist_ok=True)  # Ensure directory exists
    
    # Read existing config
    if config_file.exists():
        with open(config_file, 'r') as f:
            try:
                config = json.load(f)
            except json.JSONDecodeError:
                config = {}
    else:
        config = {}

    # Update or add new value
    config[key] = value

    # Write back all values
    with open(config_file, 'w') as f:
        json.dump(config, f, indent=2)

# Reads the value from the config file
def read_from_config(key: str) -> str | None:
    """
    Reads a value from the config file.
    Returns None if the key doesn't exist.
    """
    config_file = get_config_file()
    
    try:
        with open(config_file, 'r') as f:
            config = json.load(f)
            return config.get(key)
    except (FileNotFoundError, json.JSONDecodeError):
        return None

def get_config_dict() -> dict:
    """
    Reads the entire config file as a dictionary.
    Returns an empty dictionary if the file doesn't exist.
    """
    config_file = get_config_file()
    return json.load(open(config_file, 'r')) if config_file.exists() else {}

def replace_config(config: dict) -> None:
    """
    Replaces the entire config file with the provided dictionary.
    """
    config_file = get_config_file()
    with open(config_file, 'w') as f:
        json.dump(config, f, indent=2)

def get_customer_by_id(id: str) -> dict:
    """
    Reads the customers from the config file as a dictionary.
    """
    config = get_config_dict()
    return config.get('customers', {}).get(id, {})