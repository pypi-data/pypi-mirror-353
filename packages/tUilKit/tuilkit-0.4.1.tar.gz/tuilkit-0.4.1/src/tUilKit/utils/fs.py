# Lib/site-packages/tUilKit/utils/fs.py
"""
Contains functions for managing files, folders and path names
""" 
import shutil
import os
from tUilKit.interfaces.file_system_interface import FileSystemInterface

class FileSystem(FileSystemInterface):
    def __init__(self, logger):
        self.logger = logger

    def validate_and_create_folder(self, folder_path: str, log_file: str = None, log: bool = True, log_to: str = "both") -> bool:
        if not os.path.exists(folder_path):
            if self.logger and log:
                self.logger.colour_log("CMD", f"Attempting to create folder: {folder_path}", log_file=log_file, log_to=log_to)
            try:
                os.makedirs(folder_path, exist_ok=True)
                if self.logger and log:
                    self.logger.colour_log("DONE", f"Created folder: {folder_path}", log_file=log_file, log_to=log_to)
            except Exception as e:
                if self.logger and log:
                    self.logger.log_exception("Could not create folder: ", e, log_file=log_file, log_to=log_to)
                exit(1)
        return True

    def remove_empty_folders(self, path: str, log_file: str = None, log: bool = True) -> None:
        for root, dirs, files in os.walk(path, topdown=False):
            for dir in dirs:
                dir_path = os.path.join(root, dir)
                if not os.listdir(dir_path):
                    os.rmdir(dir_path)
                    if self.logger and log:
                        self.logger.colour_log("DONE", f"Removed empty folder: {dir_path}", log_file=log_file)

    def get_all_files(self, folder: str) -> list[str]:
        return [f for f in os.listdir(folder) if os.path.isfile(os.path.join(folder, f))]

    def validate_extension(self, fullfilepath: str, extension: str) -> str:
        base, ext = os.path.splitext(fullfilepath)
        if ext.lower() != extension.lower():
            fullfilepath += extension
        return fullfilepath

    def no_overwrite(self, fullfilepath: str, max_count=None, log_file=None, log: bool = True) -> str:
        base, ext = os.path.splitext(fullfilepath)
        counter = 1
        new_fullfilepath = fullfilepath
        oldest_file = fullfilepath
        oldest_timestamp = os.path.getmtime(fullfilepath) if os.path.exists(fullfilepath) else float('inf')
        
        while os.path.exists(new_fullfilepath):
            new_fullfilepath = f"{base}({counter}){ext}"
            if os.path.exists(new_fullfilepath):
                file_timestamp = os.path.getmtime(new_fullfilepath)
                if file_timestamp < oldest_timestamp:
                    oldest_timestamp = file_timestamp
                    oldest_file = new_fullfilepath
            counter += 1
            if max_count and counter > max_count:
                if self.logger and log:
                    self.logger.colour_log(
                        "WARN",
                        f"Max count reached, returning oldest file: {os.path.dirname(oldest_file)}/{os.path.basename(oldest_file)}",
                        log_file=log_file
                    )
                return oldest_file
        if self.logger and log:
            self.logger.colour_log(
                "DONE",
                f"No-overwrite filename generated: {os.path.dirname(new_fullfilepath)}/{os.path.basename(new_fullfilepath)}",
                log_file=log_file
            )
        return new_fullfilepath

    def backup_and_replace(self, full_pathname: str, backup_full_pathname: str, log_file=None, log: bool = True) -> str:
        if full_pathname and backup_full_pathname:
            if os.path.exists(full_pathname):
                shutil.copy2(full_pathname, backup_full_pathname)
                if self.logger and log:
                    self.logger.colour_log("DONE", f"Backup created: {backup_full_pathname}", log_file=log_file)
                try:
                    with open(full_pathname, 'w') as file:
                        file.write('')
                    if self.logger and log:
                        self.logger.colour_log("DONE", f"File replaced: {full_pathname}", log_file=log_file)
                except Exception as e:
                    if self.logger and log:
                        self.logger.log_exception("Generated Exception ", e, log_file=log_file)
        return full_pathname

    def sanitize_filename(self, filename: str) -> str:
        invalid_chars = {
            ':' : '-',
            '\\' : '',
            '/' : '',
            '?' : '',
            '*' : '',
            '<' : '',
            '>' : '',
            '|' : '',
        }
        new_filename = filename
        for char, replacement in invalid_chars.items():
            new_filename = new_filename.replace(char, replacement)
        return new_filename
