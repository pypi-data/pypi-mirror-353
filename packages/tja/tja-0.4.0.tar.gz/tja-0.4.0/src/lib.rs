mod directives;
mod parser;
mod synthesize;
mod types;

pub use directives::*;
pub use parser::*;
pub use synthesize::*;
pub use types::*;

#[cfg(feature = "python")]
mod python;
#[cfg(feature = "python")]
pub use python::*;

#[cfg(feature = "wasm")]
mod wasm;
#[cfg(feature = "wasm")]
pub use wasm::*;

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    #[test]
    fn test_parse_supernova() {
        let content = fs::read_to_string("data/SUPERNOVA.tja").unwrap();
        let mut parser = TJAParser::new();
        parser.parse_str(&content).unwrap();
        let parsed_tja = parser.get_parsed_tja();

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!("supernova", parsed_tja);
        });
    }

    #[test]
    fn test_parse_nijiiro_baton() {
        let content = fs::read_to_string("data/ニジイロバトン.tja").unwrap();
        let mut parser = TJAParser::new();
        parser.parse_str(&content).unwrap();
        let parsed_tja = parser.get_parsed_tja();
        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!("nijiiro_baton", parsed_tja);
        });
    }

    #[test]
    fn test_parse_mint_tears() {
        let content = fs::read_to_string("data/mint tears.tja").unwrap();
        let mut parser = TJAParser::new();
        parser.parse_str(&content).unwrap();
        let parsed_tja = parser.get_parsed_tja();
        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!("mint_tears", parsed_tja);
        });
    }

    #[test]
    fn test_parse_supernova_metadata_only() {
        let content = fs::read_to_string("data/SUPERNOVA.tja").unwrap();
        let mut parser = TJAParser::with_mode(ParsingMode::MetadataOnly);
        parser.parse_str(&content).unwrap();
        let parsed_tja = parser.get_parsed_tja();

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!("supernova_metadata_only", parsed_tja);
        });
    }

    #[test]
    fn test_parse_supernova_metadata_and_header() {
        let content = fs::read_to_string("data/SUPERNOVA.tja").unwrap();
        let mut parser = TJAParser::with_mode(ParsingMode::MetadataAndHeader);
        parser.parse_str(&content).unwrap();
        let parsed_tja = parser.get_parsed_tja();

        insta::with_settings!({sort_maps => true}, {
            insta::assert_json_snapshot!("supernova_metadata_and_header", parsed_tja);
        });
    }
}
