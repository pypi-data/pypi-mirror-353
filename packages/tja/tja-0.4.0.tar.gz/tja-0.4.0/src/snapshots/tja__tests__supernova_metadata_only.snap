---
source: src/lib.rs
expression: parsed_tja
snapshot_kind: text
---
{
  "metadata": {
    "bpm": 211.99990844726563,
    "offset": -1.1999999641890422,
    "demostart": 116.952,
    "songvol": 100,
    "sevol": 100,
    "raw": {
      "BPM": "211.999908447265625",
      "DEMOSTART": "116.952",
      "OFFSET": "-1.19999996418904221875",
      "SCOREMODE": "2",
      "SEVOL": "100",
      "SONGVOL": "100",
      "SUBTITLE": "--USAO",
      "TITLE": "SUPERNOVA",
      "WAVE": "SUPERNOVA.ogg"
    }
  },
  "charts": []
}
